File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/ui/screens/VibeCodeChatScreen.kt
// W10 — vibe-code session chat surface (Android parallel of iOS
// VibeCodeChatScreen). Polls GET /api/screens/llm/sessions/<id> and
// surfaces the AI's pending tool request (talkToUser or requestEnvVar)
// with a reply field. The value path for requestEnvVar POSTs first to
// /api/screens/services/<appId>/env/set (the only path that carries
// the secret), then finalizes the model-facing ack via /reply.

package com.flagshipserver.app.ui.screens

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.testTag
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import com.flagshipserver.app.api.ServiceEnvSetEnvelope
import com.flagshipserver.app.api.ServiceEnvSetRequest
import com.flagshipserver.app.api.VibeCodePendingRequest
import com.flagshipserver.app.api.VibeCodeReplyRequest
import com.flagshipserver.app.api.VibeCodeSessionPublicState
import com.flagshipserver.app.core.LocalAppState
import com.flagshipserver.app.core.LocalScreensClient
import com.flagshipserver.app.core.LocalVibeCodeEnvelopeSigner
import com.flagshipserver.app.ui.components.FSCard
import com.flagshipserver.app.ui.components.FSGhostButton
import com.flagshipserver.app.ui.components.FSPrimaryButton
import com.flagshipserver.app.ui.theme.FS
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

@Composable
fun VibeCodeChatScreen(nav: NavController, sessionId: String) {
    val client = LocalScreensClient.current
    val appState = LocalAppState.current
    val signer = LocalVibeCodeEnvelopeSigner.current
    val scope = rememberCoroutineScope()

    var state by remember { mutableStateOf<VibeCodeSessionPublicState?>(null) }
    var errorMessage by remember { mutableStateOf<String?>(null) }
    var replyDraft by remember { mutableStateOf("") }
    var envValueDraft by remember { mutableStateOf("") }
    var submitting by remember { mutableStateOf(false) }
    // Deploy state. `ready-to-deploy` (the model finished emitting files + no
    // tool pending) surfaces the Deploy button; on success the deployed URL
    // renders. The scratch deploy has no other trigger (the WS stream is a pure
    // relay), so this button IS how a chat-built service ships. Mirror of iOS.
    var deploying by remember { mutableStateOf(false) }
    var deployedUrl by remember { mutableStateOf<String?>(null) }

    val pods = appState.pods.collectAsState().value
    val leaderId = appState.leaderPodId.collectAsState().value
    val serverFqdn = pods.firstOrNull { it.podId == leaderId }?.fqdn
        ?: pods.firstOrNull()?.fqdn
        ?: "unknown"

    suspend fun reload() {
        try {
            state = client.vibeCodeSessionState(sessionId)
        } catch (t: Throwable) {
            errorMessage = t.message
        }
    }

    LaunchedEffect(sessionId) {
        reload()
        // Poll the public state while the session is still moving toward a
        // terminal state. `ready-to-deploy` is included so the screen keeps
        // refreshing until the owner taps Deploy (and reflects a deploy that
        // completed out-of-band). Mirror of iOS.
        while (true) {
            delay(1500)
            val s = state ?: continue
            if (s.status !in listOf("streaming", "awaiting-tool-response", "ready-to-deploy", "deploying")) break
            try {
                state = client.vibeCodeSessionState(sessionId)
            } catch (_: Throwable) {
                // network blip — keep polling
            }
        }
    }

    val scroll = rememberScrollState()
    Column(
        Modifier
            .fillMaxSize()
            .verticalScroll(scroll)
            .padding(horizontal = FS.space.s6),
    ) {
        Spacer(Modifier.height(FS.space.s8))
        FSGhostButton(label = "← Back", onClick = { nav.popBackStack() })
        Spacer(Modifier.height(FS.space.s3))
        Text(
            "Vibe-code session",
            color = FS.colors.text,
            style = TextStyle(fontSize = 28.sp, lineHeight = 36.sp, fontWeight = FontWeight.Medium),
        )
        Spacer(Modifier.height(FS.space.s4))

        val s = state
        if (s == null) {
            ServerCardSkeleton()
            return@Column
        }

        FSCard {
            Column {
                Text(
                    statusLabel(s.status),
                    color = FS.colors.text,
                    style = TextStyle(fontSize = 16.sp, fontWeight = FontWeight.SemiBold),
                )
                if (s.appId != null) {
                    Spacer(Modifier.height(FS.space.s1))
                    Text("app: ${s.appId}", color = FS.colors.textMuted, style = TextStyle(fontSize = 12.sp))
                }
                Spacer(Modifier.height(FS.space.s1))
                Text("session: ${s.id}", color = FS.colors.textMuted, style = TextStyle(fontSize = 12.sp))
            }
        }
        Spacer(Modifier.height(FS.space.s3))

        FSCard {
            Column {
                Text(
                    "CONVERSATION",
                    color = FS.colors.textMuted,
                    style = TextStyle(fontSize = 11.sp, fontWeight = FontWeight.SemiBold),
                )
                Spacer(Modifier.height(FS.space.s2))
                if (s.messages.isEmpty()) {
                    Text("No messages yet.", color = FS.colors.textMuted, style = TextStyle(fontSize = 13.sp))
                } else {
                    s.messages.forEach { msg ->
                        Text(
                            if (msg.role == "user") "YOU" else "AI",
                            color = FS.colors.textMuted,
                            style = TextStyle(fontSize = 11.sp, fontWeight = FontWeight.SemiBold),
                        )
                        Text(msg.text, color = FS.colors.text)
                        Spacer(Modifier.height(FS.space.s2))
                    }
                }
            }
        }
        Spacer(Modifier.height(FS.space.s3))

        val pending = s.pendingRequest
        if (pending != null) {
            when (pending) {
                is VibeCodePendingRequest.TalkToUser -> {
                    FSCard {
                        Column {
                            Text("AI asked:", color = FS.colors.textMuted, style = TextStyle(fontSize = 11.sp))
                            Spacer(Modifier.height(FS.space.s1))
                            Text(pending.payload.message, color = FS.colors.text)
                            Spacer(Modifier.height(FS.space.s3))
                            OutlinedTextField(
                                value = replyDraft,
                                onValueChange = { replyDraft = it },
                                label = { Text("Type your reply") },
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .semantics { testTag = "vibecode-reply-field" },
                            )
                            Spacer(Modifier.height(FS.space.s3))
                            FSPrimaryButton(
                                label = if (submitting) "Sending…" else "Send",
                                enabled = !submitting && replyDraft.isNotEmpty(),
                                modifier = Modifier.semantics { testTag = "vibecode-reply-send-btn" },
                                onClick = {
                                    scope.launch(Dispatchers.Main) {
                                        submitting = true
                                        try {
                                            client.vibeCodeSessionReply(
                                                sessionId,
                                                VibeCodeReplyRequest(text = replyDraft),
                                            )
                                            replyDraft = ""
                                            reload()
                                        } catch (t: Throwable) {
                                            errorMessage = t.message
                                        } finally {
                                            submitting = false
                                        }
                                    }
                                },
                            )
                        }
                    }
                }
                is VibeCodePendingRequest.RequestEnvVar -> {
                    val p = pending.payload
                    FSCard {
                        Column {
                            Text("AI needs ${p.name}", color = FS.colors.text, style = TextStyle(fontSize = 16.sp, fontWeight = FontWeight.SemiBold))
                            Spacer(Modifier.height(FS.space.s2))
                            Text(p.description, color = FS.colors.text)
                            Spacer(Modifier.height(FS.space.s1))
                            Text("Why: ${p.why}", color = FS.colors.textMuted, style = TextStyle(fontSize = 13.sp))
                            val ex = p.example
                            if (ex != null && ex.isNotEmpty()) {
                                Spacer(Modifier.height(FS.space.s1))
                                Text("Example: $ex", color = FS.colors.textMuted, style = TextStyle(fontSize = 12.sp))
                            }
                            Spacer(Modifier.height(FS.space.s3))
                            OutlinedTextField(
                                value = envValueDraft,
                                onValueChange = { envValueDraft = it },
                                label = { Text("paste your value") },
                                visualTransformation = PasswordVisualTransformation(),
                                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Password),
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .semantics { testTag = "vibecode-envvar-field" },
                            )
                            Spacer(Modifier.height(FS.space.s1))
                            Text("Sent once to your server. Not saved on your phone.", color = FS.colors.textMuted, style = TextStyle(fontSize = 11.sp))
                            Spacer(Modifier.height(FS.space.s3))
                            FSPrimaryButton(
                                label = if (submitting) "Sending…" else "Send value",
                                enabled = !submitting && envValueDraft.isNotEmpty() && s.appId != null,
                                modifier = Modifier.semantics { testTag = "vibecode-envvar-send-btn" },
                                onClick = {
                                    val appId = s.appId ?: return@FSPrimaryButton
                                    val dashIdx = appId.indexOf('-')
                                    if (dashIdx <= 0) {
                                        errorMessage = "Invalid app id shape"
                                        return@FSPrimaryButton
                                    }
                                    val creator = appId.substring(0, dashIdx)
                                    val slug = appId.substring(dashIdx + 1)
                                    scope.launch(Dispatchers.Main) {
                                        submitting = true
                                        try {
                                            val issuedAt = System.currentTimeMillis()
                                            val envelope = ServiceEnvSetEnvelope(
                                                serverId = serverFqdn,
                                                creator = creator, slug = slug,
                                                env = mapOf(p.name to envValueDraft),
                                                issuedAt = issuedAt,
                                            )
                                            val signature = signer(envelope)
                                            client.serviceEnvSet(
                                                appId,
                                                ServiceEnvSetRequest(
                                                    name = p.name, value = envValueDraft,
                                                    request = envelope, signature = signature,
                                                ),
                                            )
                                            client.vibeCodeSessionReply(
                                                sessionId,
                                                VibeCodeReplyRequest(envVarStatus = "set"),
                                            )
                                            envValueDraft = ""
                                            reload()
                                        } catch (t: Throwable) {
                                            errorMessage = t.message
                                        } finally {
                                            submitting = false
                                        }
                                    }
                                },
                            )
                            Spacer(Modifier.height(FS.space.s2))
                            FSGhostButton(
                                label = "Decline",
                                modifier = Modifier.semantics { testTag = "vibecode-envvar-decline-btn" },
                                onClick = {
                                    scope.launch(Dispatchers.Main) {
                                        try {
                                            client.vibeCodeSessionReply(
                                                sessionId,
                                                VibeCodeReplyRequest(envVarStatus = "declined"),
                                            )
                                            reload()
                                        } catch (t: Throwable) {
                                            errorMessage = t.message
                                        }
                                    }
                                },
                            )
                        }
                    }
                }
            }
            Spacer(Modifier.height(FS.space.s3))
        }

        // Deploy affordance + result. Shown once the model has finished emitting
        // the app (`ready-to-deploy`) — the only point a scratch session can be
        // shipped, since the WS stream never auto-deploys. After a successful
        // deploy the canonical URL renders. Mirror of iOS deploySection.
        val deployed = deployedUrl
        if (deployed != null || s.status == "deployed") {
            FSCard {
                Column {
                    Text(
                        "✓ Deployed",
                        color = FS.colors.success,
                        style = TextStyle(fontSize = 16.sp, fontWeight = FontWeight.SemiBold),
                    )
                    if (deployed != null) {
                        Spacer(Modifier.height(FS.space.s1))
                        Text(
                            deployed,
                            color = FS.colors.text,
                            style = TextStyle(fontSize = 13.sp),
                            modifier = Modifier.semantics { testTag = "vibecode-deployed-url" },
                        )
                    }
                }
            }
            Spacer(Modifier.height(FS.space.s3))
        } else if (s.status == "ready-to-deploy" || s.status == "deploying") {
            FSCard {
                Column {
                    Text(
                        "Ready to deploy",
                        color = FS.colors.text,
                        style = TextStyle(fontSize = 16.sp, fontWeight = FontWeight.SemiBold),
                    )
                    Spacer(Modifier.height(FS.space.s1))
                    Text(
                        "The AI finished writing your service. Deploy it to your box.",
                        color = FS.colors.textMuted,
                        style = TextStyle(fontSize = 13.sp),
                    )
                    Spacer(Modifier.height(FS.space.s3))
                    FSPrimaryButton(
                        label = if (deploying || s.status == "deploying") "Deploying…" else "Deploy",
                        enabled = !deploying && s.status == "ready-to-deploy",
                        modifier = Modifier.semantics { testTag = "vibecode-deploy-btn" },
                        onClick = {
                            scope.launch(Dispatchers.Main) {
                                deploying = true
                                try {
                                    val r = client.vibeCodeDeploy(sessionId)
                                    if (r.ok) {
                                        deployedUrl = r.url
                                    } else {
                                        errorMessage = "Deploy was rejected by the box."
                                    }
                                    reload()
                                } catch (t: Throwable) {
                                    errorMessage = t.message
                                } finally {
                                    deploying = false
                                }
                            }
                        },
                    )
                }
            }
            Spacer(Modifier.height(FS.space.s3))
        }

        if (errorMessage != null) {
            ErrorCard(message = errorMessage!!)
        }
        Spacer(Modifier.height(FS.space.s10))
    }
}

private fun statusLabel(status: String): String = when (status) {
    "streaming" -> "Generating…"
    "awaiting-tool-response" -> "AI is asking you something"
    "ready-to-deploy" -> "Ready to deploy"
    "deploying" -> "Deploying…"
    "deployed" -> "Deployed"
    "failed" -> "Failed"
    "cancelled" -> "Cancelled"
    else -> status
}

File: apps/mobile/ios/Sources/FlagshipUI/Screens/VibeCodeChatScreen.swift
import SwiftUI
import FlagshipAPI
import FlagshipCore

/// W10 — vibe-code session chat surface.
///
/// Subscribes to GET /api/screens/llm/sessions/<id> (poll-driven) and
/// renders the running message log. When the session is
/// `awaiting-tool-response`, surfaces the model's tool request:
///
///   - talkToUser    — single text field for a free-form reply,
///                     POSTed to /reply.
///   - requestEnvVar — single field labeled with the env-var NAME the
///                     model wants. On send, the value is POSTed to
///                     /api/screens/services/<appId>/env/set FIRST
///                     (the path that actually carries the secret),
///                     then /reply finalizes the ack with status "set".
///
/// The push handler at the app delegate level routes the
/// `vibecode-needs-you` payload here via deep link
/// `flagship://vibecode/<sessionId>`.
@MainActor
public struct VibeCodeChatScreen: View {
    @Environment(\.colorScheme) private var scheme
    @State private var state: VibeCodeSessionPublicState?
    @State private var loading: Bool = true
    @State private var errorMessage: String?
    @State private var replyDraft: String = ""
    @State private var envValueDraft: String = ""
    @State private var submitting: Bool = false
    @State private var pollTask: Task<Void, Never>?
    /// Deploy state. `ready-to-deploy` (the model finished emitting files +
    /// no tool pending) surfaces the Deploy button; on success the deployed
    /// URL renders. The scratch deploy has no other trigger (the WS stream is
    /// a pure relay), so this button IS how a chat-built service ships.
    @State private var deploying: Bool = false
    @State private var deployedUrl: String?
    @State private var deployedServiceId: String?

    public let sessionId: String
    public let serverFqdn: String
    public let username: String
    public let client: any ScreensClient
    /// Same hook as ServiceEnvScreen — signs a SetServiceEnvRequest
    /// envelope under the owner's IRK. Required because requestEnvVar
    /// acks may need to first push the secret value through /env/set.
    public let signEnvelope: @MainActor (ServiceEnvSetEnvelope) async throws -> String

    public init(
        sessionId: String,
        serverFqdn: String,
        username: String,
        client: any ScreensClient,
        signEnvelope: @escaping @MainActor (ServiceEnvSetEnvelope) async throws -> String
    ) {
        self.sessionId = sessionId
        self.serverFqdn = serverFqdn
        self.username = username
        self.client = client
        self.signEnvelope = signEnvelope
    }

    public var body: some View {
        let c = FSColors.scheme(scheme)
        ScrollView {
            VStack(alignment: .leading, spacing: FS.space.s4) {
                if let s = state {
                    statusCard(state: s, c: c)
                    messagesCard(state: s, c: c)
                    if let pending = s.pendingRequest {
                        replySection(state: s, pending: pending, c: c)
                    }
                    deploySection(state: s, c: c)
                } else if loading {
                    FSCard {
                        HStack {
                            ProgressView()
                            Text("Loading session…").foregroundColor(c.textMuted)
                        }
                    }
                }
                if let msg = errorMessage {
                    ErrorCard(message: msg)
                }
                Spacer().frame(height: FS.space.s12)
            }
            .padding(.horizontal, FS.space.s6)
            .padding(.top, FS.space.s4)
        }
        .scrollDismissesKeyboard(.interactively)
        .background(c.bg.ignoresSafeArea())
        .navigationTitle("Vibe-code session")
        .navigationBarTitleDisplayMode(.inline)
        .task {
            await reload()
            startPolling()
        }
        .onDisappear {
            pollTask?.cancel()
            pollTask = nil
        }
    }

    /// Deploy affordance + result. Shown once the model has finished emitting
    /// the app (`ready-to-deploy`) — the only point a scratch session can be
    /// shipped, since the WS stream never auto-deploys. After a successful
    /// deploy the canonical URL renders with an Open affordance.
    @ViewBuilder
    private func deploySection(state s: VibeCodeSessionPublicState, c: FSColors) -> some View {
        if let url = deployedUrl ?? (s.status == "deployed" ? "" : nil), !(deployedUrl == nil && url.isEmpty) {
            FSCard {
                VStack(alignment: .leading, spacing: FS.space.s2) {
                    HStack(spacing: FS.space.s2) {
                        Image(systemName: "checkmark.seal.fill").foregroundColor(c.success)
                        Text("Deployed").font(FS.font.h4()).foregroundColor(c.text)
                    }
                    Text(url).font(FS.font.mono()).foregroundColor(c.text)
                        .lineLimit(1).truncationMode(.middle)
                        .accessibilityIdentifier("vibecode-deployed-url")
                }
            }
        } else if s.status == "ready-to-deploy" || s.status == "deploying" {
            FSCard {
                VStack(alignment: .leading, spacing: FS.space.s3) {
                    Text("Ready to deploy").font(FS.font.h4()).foregroundColor(c.text)
                    Text("The AI finished writing your service. Deploy it to your box.")
                        .font(FS.font.bodySm()).foregroundColor(c.textMuted)
                    FSPrimaryButton(
                        (deploying || s.status == "deploying") ? "Deploying…" : "Deploy",
                        enabled: !deploying && s.status == "ready-to-deploy",
                        block: true
                    ) {
                        Task { await deploy() }
                    }
                    .accessibilityIdentifier("vibecode-deploy-btn")
                }
            }
        }
    }

    private func deploy() async {
        deploying = true
        defer { deploying = false }
        do {
            let r = try await client.vibeCodeDeploy(sessionId: sessionId)
            if r.ok {
                deployedUrl = r.url
                deployedServiceId = r.serviceId
            } else {
                errorMessage = "Deploy was rejected by the box."
            }
            await reload()
        } catch {
            errorMessage = ScreensClientError.userFacing(error)
        }
    }

    private func statusCard(state s: VibeCodeSessionPublicState, c: FSColors) -> some View {
        FSCard {
            VStack(alignment: .leading, spacing: FS.space.s2) {
                HStack(spacing: FS.space.s2) {
                    Image(systemName: statusIcon(s.status)).foregroundColor(statusColor(s.status, c: c))
                    Text(statusLabel(s.status))
                        .font(FS.font.h4())
                        .foregroundColor(c.text)
                }
                if let appId = s.appId {
                    Text("app: \(appId)").font(FS.font.mono()).foregroundColor(c.textMuted)
                }
                Text("session: \(s.id)").font(FS.font.mono()).foregroundColor(c.textMuted)
            }
        }
    }

    private func statusIcon(_ status: String) -> String {
        switch status {
        case "streaming": return "wand.and.stars"
        case "awaiting-tool-response": return "questionmark.bubble.fill"
        case "ready-to-deploy": return "checkmark.circle.fill"
        case "deployed": return "checkmark.seal.fill"
        case "failed", "cancelled": return "xmark.octagon.fill"
        default: return "circle"
        }
    }

    private func statusColor(_ status: String, c: FSColors) -> Color {
        switch status {
        case "awaiting-tool-response": return c.primary
        case "deployed": return c.success
        case "failed", "cancelled": return c.danger
        default: return c.textMuted
        }
    }

    private func statusLabel(_ status: String) -> String {
        switch status {
        case "streaming": return "Generating…"
        case "awaiting-tool-response": return "AI is asking you something"
        case "ready-to-deploy": return "Ready to deploy"
        case "deploying": return "Deploying…"
        case "deployed": return "Deployed"
        case "failed": return "Failed"
        case "cancelled": return "Cancelled"
        default: return status
        }
    }

    private func messagesCard(state s: VibeCodeSessionPublicState, c: FSColors) -> some View {
        FSCard {
            VStack(alignment: .leading, spacing: FS.space.s3) {
                Text("CONVERSATION").font(FS.font.caption()).tracking(1).foregroundColor(c.textMuted)
                if s.messages.isEmpty {
                    Text("No messages yet.").font(FS.font.bodySm()).foregroundColor(c.textMuted)
                } else {
                    ForEach(Array(s.messages.enumerated()), id: \.offset) { _, msg in
                        messageRow(msg: msg, c: c)
                    }
                }
            }
        }
    }

    private func messageRow(msg: VibeCodeSessionMessage, c: FSColors) -> some View {
        VStack(alignment: .leading, spacing: FS.space.s1) {
            Text(msg.role == "user" ? "YOU" : "AI")
                .font(.system(size: 11, weight: .semibold)).tracking(1)
                .foregroundColor(c.textMuted)
            Text(msg.text)
                .font(FS.font.body())
                .foregroundColor(c.text)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(.vertical, 4)
    }

    @ViewBuilder
    private func replySection(
        state s: VibeCodeSessionPublicState,
        pending: VibeCodePendingRequest,
        c: FSColors
    ) -> some View {
        switch pending {
        case .talkToUser(_, let message):
            talkToUserCard(message: message, c: c)
        case .requestEnvVar(_, let name, let description, let why, let example, let secret):
            requestEnvVarCard(
                state: s,
                pending: pending,
                name: name,
                description: description,
                why: why,
                example: example,
                secret: secret ?? false,
                c: c
            )
        }
    }

    private func talkToUserCard(message: String, c: FSColors) -> some View {
        FSCard {
            VStack(alignment: .leading, spacing: FS.space.s3) {
                Text("AI asked:").font(FS.font.caption()).tracking(1).foregroundColor(c.textMuted)
                Text(message).font(FS.font.body()).foregroundColor(c.text)
                TextField("Type your reply", text: $replyDraft, axis: .vertical)
                    .lineLimit(3...6)
                    .padding(10)
                    .background(c.surfaceSunken)
                    .overlay(RoundedRectangle(cornerRadius: FS.radius.sm).stroke(c.border, lineWidth: 1))
                    .clipShape(RoundedRectangle(cornerRadius: FS.radius.sm))
                    .accessibilityIdentifier("vibecode-reply-field")
                FSPrimaryButton(
                    submitting ? "Sending…" : "Send",
                    enabled: !submitting && !replyDraft.isEmpty,
                    block: true
                ) {
                    Task { await submitReply() }
                }
                .accessibilityIdentifier("vibecode-reply-send-btn")
            }
        }
    }

    private func requestEnvVarCard(
        state s: VibeCodeSessionPublicState,
        pending: VibeCodePendingRequest,
        name: String,
        description: String,
        why: String,
        example: String?,
        secret: Bool,
        c: FSColors
    ) -> some View {
        FSCard {
            VStack(alignment: .leading, spacing: FS.space.s3) {
                HStack(spacing: FS.space.s2) {
                    Image(systemName: secret ? "lock.shield" : "key").foregroundColor(c.primary)
                    Text("AI needs \(name)").font(FS.font.h4()).foregroundColor(c.text)
                }
                Text(description).font(FS.font.body()).foregroundColor(c.text)
                Text("Why: \(why)").font(FS.font.bodySm()).foregroundColor(c.textMuted)
                if let ex = example, !ex.isEmpty {
                    Text("Example: \(ex)").font(FS.font.mono()).foregroundColor(c.textMuted)
                }
                SecureField("paste your value here", text: $envValueDraft)
                    .font(FS.font.mono())
                    .autocorrectionDisabled()
                    .textInputAutocapitalization(.never)
                    .padding(.horizontal, 12).frame(height: 40)
                    .background(c.surfaceSunken)
                    .overlay(RoundedRectangle(cornerRadius: FS.radius.sm).stroke(c.border, lineWidth: 1))
                    .clipShape(RoundedRectangle(cornerRadius: FS.radius.sm))
                    .accessibilityIdentifier("vibecode-envvar-field")
                Text("Sent once to your server. Not saved on your phone.")
                    .font(FS.font.caption()).foregroundColor(c.textMuted)
                HStack(spacing: FS.space.s2) {
                    FSPrimaryButton(
                        submitting ? "Sending…" : "Send value",
                        enabled: !submitting && !envValueDraft.isEmpty && s.appId != nil,
                        block: true
                    ) {
                        Task { await submitEnvVar(state: s, name: name) }
                    }
                    .accessibilityIdentifier("vibecode-envvar-send-btn")
                }
                Button("Decline") {
                    Task { await declineEnvVar() }
                }
                .font(FS.font.bodySm())
                .foregroundColor(c.textMuted)
                .accessibilityIdentifier("vibecode-envvar-decline-btn")
            }
        }
    }

    private func reload() async {
        loading = true
        errorMessage = nil
        defer { loading = false }
        do {
            state = try await client.vibeCodeSessionState(sessionId: sessionId)
        } catch {
            errorMessage = ScreensClientError.userFacing(error)
        }
    }

    private func startPolling() {
        pollTask?.cancel()
        pollTask = Task { @MainActor in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 1_500_000_000)
                if Task.isCancelled { break }
                // Poll while the session is still moving toward a terminal
                // state. `ready-to-deploy` is included so the screen keeps
                // refreshing until the owner taps Deploy (and reflects a
                // deploy that completed out-of-band).
                if let s = state, ["streaming", "awaiting-tool-response", "ready-to-deploy", "deploying"].contains(s.status) {
                    do {
                        state = try await client.vibeCodeSessionState(sessionId: sessionId)
                    } catch {
                        // network blip — keep polling
                    }
                }
            }
        }
    }

    private func submitReply() async {
        submitting = true
        defer { submitting = false }
        do {
            let _ = try await client.vibeCodeSessionReply(
                sessionId: sessionId,
                VibeCodeReplyRequest(text: replyDraft, envVarStatus: nil)
            )
            replyDraft = ""
            await reload()
        } catch {
            errorMessage = ScreensClientError.userFacing(error)
        }
    }

    private func submitEnvVar(state s: VibeCodeSessionPublicState, name: String) async {
        guard let appId = s.appId else {
            errorMessage = "Session has no app id yet — wait for the manifest to emit."
            return
        }
        // Derive (creator, slug) from the appId. appId = "<creator>--<slug>";
        // split on the `--` delimiter (both halves may carry single dashes —
        // docs/service-addressing-double-dash.md).
        guard let delim = appId.range(of: "--") else {
            errorMessage = "Invalid app id shape"
            return
        }
        let creator = String(appId[..<delim.lowerBound])
        let slug = String(appId[delim.upperBound...])
        submitting = true
        defer { submitting = false }
        let issuedAt = Int64(Date().timeIntervalSince1970 * 1000)
        let envelope = ServiceEnvSetEnvelope(
            serverId: serverFqdn,
            creator: creator,
            slug: slug,
            env: [name: envValueDraft],
            issuedAt: issuedAt
        )
        do {
            // Step 1: push the secret value through /env/set (the only
            // path that carries it).
            let signature = try await signEnvelope(envelope)
            let _ = try await client.serviceEnvSet(
                appId: appId,
                ServiceEnvSetRequest(name: name, value: envValueDraft, request: envelope, signature: signature)
            )
            // Step 2: finalize the model-facing ack via /reply.
            let _ = try await client.vibeCodeSessionReply(
                sessionId: sessionId,
                VibeCodeReplyRequest(text: nil, envVarStatus: "set")
            )
            envValueDraft = ""
            await reload()
        } catch {
            errorMessage = ScreensClientError.userFacing(error)
        }
    }

    private func declineEnvVar() async {
        submitting = true
        defer { submitting = false }
        do {
            let _ = try await client.vibeCodeSessionReply(
                sessionId: sessionId,
                VibeCodeReplyRequest(text: nil, envVarStatus: "declined")
            )
            await reload()
        } catch {
            errorMessage = ScreensClientError.userFacing(error)
        }
    }
}
